hwcomposer/hwcomposer_backend_v20.cpp

```diff
@@ -195,25 +203,14 @@ void HWC2Window::present(HWComposerNativeWindowBuffer *buffer)
 
     QPA_HWC_TIMING_SAMPLE(setTime);
 
-    hwc2_compat_out_fences_t* fences;
-    error = hwc2_compat_display_get_release_fences(
-        hwcDisplay, &fences);
-
-    if (error != HWC2_ERROR_NONE) {
-        qDebug("presentAndGetReleaseFences: Failed to get release fences "
-              "for display %d: %d", displayId, error);
-        return;
+    if (lastPresentFence != -1) {
+        sync_wait(lastPresentFence, -1);
+        close(lastPresentFence);
     }
 
-    int fenceFd = hwc2_compat_out_fences_get_fence(fences, layer);
-    if (fenceFd != -1)
-        setFenceBufferFd(buffer, fenceFd);
-    else if (presentFence != -1)
-        setFenceBufferFd(buffer, presentFence);
-
-    hwc2_compat_out_fences_destroy(fences);
+    lastPresentFence = presentFence != -1 ? dup(presentFence) : -1;
```

Comment:
Since this seems to be an issue specific to some hardware and this is a work-around of sorts maybe this should be conditionally enabled with an environment variable.

Reply:
does it have a negative impact on other devices?

Reply:
I do not know. I'm leaning towards no and I think eglSwapBuffer may actually be supposed to block in this manner but my thinking might be specific to the threaded render loop use case.

Reply:
I've compared two L500Ds running side by side with the equivalent change to the v11 backend which I've appended as an additional patch and with https://git.merproject.org/mer-core/qtwayland/merge_requests/15/diffs and the results were inconclusive at least to begin with.

The application frame rate in the settings main page was initially poor with the change, which I think I see sometimes anyway because the rhythm between the compositor and application are out of sync and the compositor steals all the GPU resources. Adding QPA_HWC_IDLE_TIME=10 helps with that and I think we were supposed to be setting a value for that for all devices anyway after which the two devices seemed on par, maintaining a not quite 60fps, after being left to run unattended for a while the unmodified device seems to have settled at a very constant 60fps.

The frame rate display from developer tools isn't an ideal assessment though as it tendency to change the result through measuring it because it forces the compositor to constantly redraw instead of after the application window updates like in the normal case, so not setting QPA_HWC_IDLE_TIME may not be as important for real use.